### ISSUE TITLE
Should onException catch?

### ISSUE THREAD

**Issue by despresc** [Gabriel439/Haskell-Pipes-Safe-Library#69]
I pointed out in [this reddit comment](https://old.reddit.com/r/haskell/comments/1f35f92/how_is_pipessafe_supposed_to_work/lkc1hsf/) that the `pipes-safe` version of `onException` is implemented with `register`/`release`:

https://github.com/Gabriella439/Haskell-Pipes-Safe-Library/blob/b76b3b38500da8c1c9e41525a5bf453ccd50a12b/src/Pipes/Safe.hs#L412-L417

but [the version in Control.Exception](https://hackage.haskell.org/package/base-4.20.0.1/docs/Control-Exception.html#v:onException) is implemented with `catch`. This leads to the surprising behaviour noted in that thread; in the `base` version of `finally`, the final action will be run before the exception is propagated upward, whereas in the `pipes-safe` version the final action is only guaranteed to be run when the `MonadSafe` scope is exited.

I haven't tested this yet, but could this library's version of `onException` be changed to something like:

```
onException :: (MonadSafe m) => m a -> Base m b -> m a
m1 `onException` io = do
    key <- register (io >> return ())
    r   <- m1 `catch` (\e -> release key >> throwM (e :: SomeException))
    release key
    return r
```

so `finally` and related `pipes-safe` functions will behave more like their `base` counterparts?

Also, the documentation of `Control.Exception.onException` claims that it "only performs the final action if there was an exception raised by the computation", but it looks like the `pipes-safe` version will always run the final action (which is how `Control.Exception.finally` behaves). The documentation should probably mention that if that is intentional.

**Comment by despresc** [comment]
Right after posting this I found [this comment](https://groups.google.com/g/haskell-pipes/c/ZZjCAEDqrdc/m/Rxs4nJYM3BEJ), so I'm assuming this (the not-so-prompt finalization) is indeed a bug.

**Comment by tomjaguarpaw** [comment]
Your suggested version looks pretty plausible. I think it's worth trying. However, suppose it's applied to a pipe to obtain `p1`, and we compose `p1` with `p2` like `p1 >-> p2`. I'm skeptical that resources in `p2` wil be freed promptly when there's an exception in `p1`. I think it should be tried though!

**Comment by despresc** [comment]
I agree that my version won't have the behaviour you want in the `p1 >-> p2` scenario.

**Comment by Gabriella439** [comment]
It's been a loong time before I thought about all of this, but this *sounds* okay to me.  I'm definitely open to reviewing any PR someone puts up related to this and it might help jog my memory on some of this logic.
